torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -22,73 +22,93 @@ AccumulateGrad::AccumulateGrad(Variable variable_)
     // slightly more efficient.
-    Variable& grad_variable = as_variable_ref(grad);
-    if (grad_variable.is_sparse() && !new_grad.is_sparse()) {
-      // If `grad_variable` is sparse and `new_grad` is not sparse, their sum is not
-      // sparse, and we must change the TensorImpl type of `grad_variable` for it to
-      // store the result. However, changing the TensorImpl type of a tensor requires
-      // changing the tensor itself, and thus in this case we have to change the grad
-      // tensor.
-      grad_variable = new_grad + grad_variable;
+    if (variable_grad.is_sparse() && !new_grad_copy.is_sparse()) {
+      // If `grad_variable` is sparse and `new_grad_copy` is not sparse, their
+      // sum is not sparse, and we must change the TensorImpl type of
+      // `grad_variable` for it to store the result. However, changing the
+      // TensorImpl type of a tensor requires changing the tensor itself, and
+      // thus in this case we have to change the grad tensor.
+      update_grad_fn(new_grad_copy + variable_grad);
     } else {
-      // In this case we can avoid changing the grad tensor. There are three scenarios
-      // when we'll hit this case:
+      // In this case we can avoid changing the grad tensor. There are three
+      // scenarios when we'll hit this case:
       //
-      // 1. `grad_variable` is sparse, and `new_grad` is sparse.
-      // 2. `grad_variable` is dense, and `new_grad` is sparse.
-      // 3. `grad_variable` is dense, and `new_grad` is dense.
+      // 1. `grad_variable` is sparse, and `new_grad_copy` is sparse.
+      // 2. `grad_variable` is dense, and `new_grad_copy` is sparse.
+      // 3. `grad_variable` is dense, and `new_grad_copy` is dense.
       //
-      // In all of these three cases, `grad_variable += new_grad` is a valid operation
-      // which adds `new_grad` to `grad_variable` in place. `grad_variable` is thus
-      // still referring to the same tensor after the operation.
-      grad_variable += new_grad;
+      // In all of these three cases, `grad_variable += new_grad_copy` is a
+      // valid operation which adds `new_grad_copy` to `grad_variable` in place.
+      // `grad_variable` is thus still referring to the same tensor after the
+      // operation.
+      variable_grad += new_grad_copy;
     }
   } else {
-    variable.grad() = grad + new_grad;
+    update_grad_fn(variable_grad + new_grad_copy);
   }
+}
+
+auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
+  // XXX: this method is not thread-safe!
+  check_input_variables("AccumulateGrad", grads, 1, 0);
+
+  if (!grads[0].defined())
+    return {};
+  if (variable.grad_fn())
+    throw std::logic_error(
+        "leaf variable has been moved into the graph interior");
+  if (!variable.requires_grad())
+    return {};
+
+  auto new_grad = std::move(grads[0]);
+  at::Tensor& grad = variable.grad();
+  accumulateGradAndCallHooks(
+      variable,
+      grad,
+      new_grad,
```

Comment:
A little confused why you don't `std::move(new_grad)` here

Reply:
Obviously the `std::move` does nothing if `accumulateGradAndCallHooks` takes by reference, but I don't see any good reason for it to take the arg by reference.

Reply:
The reason `accumulateGradAndCallHooks` takes a const reference is because we would like to avoid a copy that would increment the Tensor's refcount. We can use std::move() here to avoid the refcount bump, although we can't use std::move() when we call `accumulateGradAndCallHooks` since the grad there is a const reference that is owned by the Future and the Future only allows callbacks to operate on a const ref of the value (which makes sense for the Future API).

Reply:
Let me put it differently. Why do you move the tensor out of the passed in `variables_list`? If you're going to pass the tensor on by reference (as happens here), then there's no point in the move. Basically, the code is inconsistent: if it moves, it should move all the way to a point where the moves pay off (called function takes ownership); otherwise, you should just pass by reference all around. It sounds like you're in the latter situation.

Reply:
I have rereviewed the code after your changes and I think the original move was justified. See my comments above. You should put back the move here.

Reply:
The additional cost for using const ref is we need to make a shallow copy inside accumulateGradAndCallHooks.  Given the current code structure, this seems to be reasonable to me. In the future, when we move dist grad accumulation to be inline (just like local grad), we can change this to std::move.